core/cap-0007.md

```diff
@@ -0,0 +1,150 @@
+## Preamble
+
+```
+CAP: 0007
+Title: Deterministic Account Creation
+Author: Jeremy Rubin
+Status: Draft
+Created: 2018-11-07
+Discussion: https://github.com/stellar/stellar-protocol/issues/99
+Protocol version: TBD
+```
+
+## Simple Summary
+We introduce CreateDeterministicAccount operation which creates an account that
+could only be created by the transaction creating it, allowing it to have a
+known starting sequence number and initialization.
+
+## Abstract
+
+
+Deterministic accounts are a critical infrastructure piece for building reliable contracts for three key reasons:
+
+1) Predictable sequence numbers
+2) Never-fail creation
+3) Known initialized state
+
+We can create deterministic accounts with minimal changes to Stellar by using Key Tweaking to derive the account key.
+
+
+## Motivation
+
+Currently, when a contract must create an account as a part of it's settlement (such as in Starlight or in an Account Transfer),
+creating the account results in an unpredictable sequence number which prevents the presigning of transactions for it. Furthermore,
+accounts can not be created and modified at the same time preventing initializing an account on behalf of another party.
+
+
+## Specification
+
+We assume a function `PublicKey derive(PublicKey pk, const *char buf, size_t
+buflen)` as specified by [ChainKD Non Hardened Extended Public
+Key](https://chain.com/docs/1.2/protocol/specifications/chainkd#derive-non-hardened-extended-public-key) which returns
+keys of type `PUBLIC_KEY_TYPE_ED25519_DET`.
+
+We extend the Key types:
+
+```c++
+
+enum CryptoKeyType
+{
+    KEY_TYPE_ED25519 = 0,
+    KEY_TYPE_PRE_AUTH_TX = 1,
+    KEY_TYPE_HASH_X = 2
+    KEY_TYPE_ED25519_DET = 3
+};
+
+enum PublicKeyType
+{
+    PUBLIC_KEY_TYPE_ED25519 = KEY_TYPE_ED25519,
+    PUBLIC_KEY_TYPE_ED25519_DET = KEY_TYPE_ED25519_DET,
+};
+
+
+union PublicKey switch (PublicKeyType type)
+{
+case PUBLIC_KEY_TYPE_ED25519:
+    uint256 ed25519;
+case PUBLIC_KEY_TYPE_ED25519_DET:
+	uint256 ed25519;
+};
+
+
+```
+
+
+
+We add the following operation:
+```c++
+struct CreateDeterministicAccountOp
+{
+    PublicKey keyToDeriveFrom;
+    int64 startingBalance;
+    *Signer first_signer;
+	uint32 masterWeight;
+
+};
+```
+
+The result of this operation is an account created with the following state:
+```c++
+AccountEntry
+{
+    accountID = derive(op.keyToDeriveFrom, SHA512()
+										.hash(tx.hash)
+										.hash(op.index)
+										.digest() ,64),
+    balance = op.startingBalance,
+    seqNum = 0,
```

Comment:
Sequence number 0 have a special signification in SEP-0007 where it means "replace me by current sequence number". I don't see any real incompatibility, but maybe this is better to use seqNum=1 anyway as sequence=0 were not supposed to be possible until here?

Reply:
... or not use magic numbers in the first place? I thought that idea was abandoned in favor of templates as it's brittle and a recipe for bad code in the wild? In the future we may also want to use the full range of sequence numbers.

Reply:
@MisterTicot if the seq num starts at zero, then the next transaction is at seqNum 1, so there is no incompatibility.

We will never execute a txn at sequence number 0.